syntax = "proto3";

package course;

service CourseService {
    rpc GetCourse           (GetCourseRequest)  returns (GetCourseResponse);
    rpc ListCourses         (ListCoursesRequest)    returns (ListCoursesResponse);
    rpc ListCoursesByIds    (ListCoursesByIdsRequest)   returns (ListCoursesByIdsResponse);
    rpc GetCourseLessons   (GetCourseLessonsRequest)   returns (GetCourseLessonsResponse);
}


message GetCourseRequest {
    string id = 1;
}

message CoursePrice {
    string type = 1;
    int32 amount = 2;
}

message CourseSection {
    string title = 1;
    repeated string items = 2;
}

message GetCourseResponse {
    string id = 1;
    string title = 2;
    string description = 3;

    string level = 4;
    int32 duration_minutes = 5;
    string language = 6;
    bool requires_experience = 7;

    CoursePrice price = 8;

    repeated string category_ids = 9;
    repeated string mentor_ids = 10;
    string cover_image_id = 11;

    string author_id = 12;
    int64 lessons_count = 13;

    repeated CourseSection sections = 14;
}


message ListCoursesRequest {
    int32 limit  = 1;
    int32 offset = 2;
    string mode = 3;
    string author_id = 4;
}

message Course {
    string id = 1;
    string title = 2;
    string description = 3;

    int32 duration_minutes = 4;
    string language = 5;

    CoursePrice price = 6;

    repeated string category_ids = 7;
    string cover_image_id = 8;

    bool is_promoted = 9;
}

message ListCoursesResponse {
    repeated Course items = 1;
    int32 total = 2;
}

message ListCoursesByIdsRequest {
    repeated string ids = 1;
}

message ListCoursesByIdsResponse {
    repeated Course items = 1;
}

message GetCourseLessonsRequest {
    string course_id = 1;
}

message CourseLesson {
    string id = 1;
    string title = 2;

    string type = 3;
    string access_type = 4;
    string duration = 5;

    int32 order_number = 6;
    bool is_active = 7;
}

message GetCourseLessonsResponse {
    repeated CourseLesson items = 1;
}